Add support for Linux / ARM platform (Raspberry Pi)

lispmeister opened the pull request:
To build on the Raspberry Pi we need additional flags in the
Makefile and in emitglue.c.

Add: Makefile conditionals and GCC conditionals

lispmeister committed:
Add support for Linux / ARM platform (Raspberry Pi)

To build on the Raspberry Pi we need additional flags in the
Makefile and in emitglue.c.

Add: Makefile conditionals and GCC conditionals

dpgeorge commented on unix/Makefile:
```diff
@@ -74,10 +77,14 @@ ifeq ($(MICROPY_FORCE_32BIT),1)
 ifeq ($(MICROPY_PY_FFI),1)
 ifeq ($(UNAME_S),Linux)
+ifeq ($(UNAME_M), armv7l)
+CFLAGS_MOD += -I/usr/include/arm-linux-gnueabihf
```
I don't understand why this is needed.  Why do you use MICROPY_FORCE_32BIT=1?

lispmeister replied:
I don't. This is just in case you need to set the bit and are building on an ARM platform.

dpgeorge replied:
You mean an ARM 64-bit platform?  But isn't that only since ARMv8?

lispmeister replied:
The Raspberry Pi 3 just came out. It would be detected as a 64-bit platform and the build would break.
I any case, if you think it's confusing, I'll remove it from this PR.

dpgeorge replied:
Did you try your patch on a 64-bit ARM machine?

MICROPY_FORCE_32BIT=1 should only be needed for 64-bit machines (and should be harmless on 32-bit machines).

This block of code here in the Makefile is specific to the case where you're on a 64-bit machine, you used MICROPY_FORCE_32BIT=1, and you also want modffi.  In the case of a 32-bit RPi, it should work without your addition.  In the case of a 64-bit RPi, I don't know what happens, but I would have thought that, if anything, you'd want to have `if ($(UNAME_M),armv8)`.

pfalcon commented:
MicroPython is known to build for RaspPi without any such changes. So, please first investigate why other people didn't need it, while you need, and include them as the comment to pull request.

dpgeorge commented:
> So, please first investigate why other people didn't need it,

I think my persistent bytecode changes to emitglue.c were made after RPi testing, hence the need for that particular fix.

lispmeister commented:
Without the two patched lines in emitglue.c the unix file reader isn't available on the Raspberry Pi. I'll remove the controversial Makefile changes from this PR. I won't be able to confirm they work as intended on a 64-bit ARM platform (Raspberry Pi 3) this week. I'll submit a new PR once that has been tested.

lispmeister committed:
Undo 32-bit customization for ARM platform in Makefile

We can't test the change on a 64-bit ARM platform to ensure it
works. Better to remove it until the change can be verified.

Mod: Undo change in Makefile

dpgeorge commented:
Merged in 4f0080346b565cbf9f2204874b81814013d67488, thanks!

dhylands commented:
Even though the RPi 3 is a 64-bit ARM, the kernel and userspace are all currently 32-bit, so fundamentally the same as the RPI 2.

Until they release a 64-bit kernel, then the RPI 3 will look/feel like an RPi 2 as for a 32 versus 64-bit is concerned.
